File: Frontend/src/app/services/ad.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface Ad {
  id: number;
  type: AdType;
  content?: string;
  url?: string;
  thumbnailUrl?: string;
  costPerClick: number;
  costPerView: number;
}

export interface AdCampaign {
  id: number;
  name: string;
  description?: string;
  advertiserId: string;
  type: AdType;
  budget: number;
  costPerClick: number;
  costPerView: number;
  startDate: string;
  endDate: string;
  status: AdStatus | string;
  targetAudience?: string;
  adContent?: string;
  adUrl?: string;
  thumbnailUrl?: string;
  createdAt: string;
  updatedAt?: string;
  advertiser?: any;
  impressions?: AdImpression[];
}

export interface AdCampaignCreate {
  name: string;
  description?: string;
  type: AdType;
  budget: number;
  costPerClick: number;
  costPerView: number;
  startDate: string;
  endDate: string;
  targetAudience?: string;
  adContent?: string;
  adUrl?: string;
  thumbnailUrl?: string;
}

export interface AdCampaignUpdate {
  name?: string;
  description?: string;
  type?: AdType;
  budget?: number;
  costPerClick?: number;
  costPerView?: number;
  startDate?: string;
  endDate?: string;
  status?: AdStatus;
  targetAudience?: string;
  adContent?: string;
  adUrl?: string;
  thumbnailUrl?: string;
}

export interface AdImpression {
  id: number;
  adCampaignId: number;
  videoId?: number;
  userId?: string;
  viewedAt: string;
  wasClicked: boolean;
  clickedAt?: string;
  revenue: number;
  user?: any;
  video?: any;
}

export interface AdRevenue {
  totalRevenue: number;
  totalImpressions: number;
  totalClicks: number;
  clickThroughRate: number;
  revenuePerImpression: number;
  revenuePerClick: number;
}

export interface AdAnalytics {
  campaignId: number;
  totalImpressions: number;
  totalClicks: number;
  totalRevenue: number;
  clickThroughRate: number;
  costPerClick: number;
  costPerImpression: number;
}

export enum AdType {
  Banner = 0,
  Video = 1,
  Overlay = 2,
  Sponsored = 3
}

export enum AdStatus {
  Draft = 0,
  Active = 1,
  Paused = 2,
  Completed = 3,
  Cancelled = 4
}

@Injectable({
  providedIn: 'root'
})
export class AdService {
  private apiUrl = `${environment.apiUrl}/api/ad`;

  constructor(private http: HttpClient) {}

  // Campaign Management
  createCampaign(campaign: AdCampaignCreate): Observable<AdCampaign> {
    return this.http.post<AdCampaign>(`${this.apiUrl}/campaigns`, campaign);
  }

  getCampaign(id: number): Observable<AdCampaign> {
    return this.http.get<AdCampaign>(`${this.apiUrl}/campaigns/${id}`);
  }

  getCampaigns(page: number = 1, pageSize: number = 20): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/campaigns?page=${page}&pageSize=${pageSize}`);
  }

  updateCampaign(id: number, campaign: AdCampaignUpdate): Observable<AdCampaign> {
    return this.http.put<AdCampaign>(`${this.apiUrl}/campaigns/${id}`, campaign);
  }

  deleteCampaign(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/campaigns/${id}`);
  }

  activateCampaign(id: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/campaigns/${id}/activate`, {});
  }

  pauseCampaign(id: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/campaigns/${id}/pause`, {});
  }

  // Ad Serving
  getAd(videoId?: number, type?: AdType): Observable<Ad | null> {
    let params = '';
    if (videoId) params += `videoId=${videoId}`;
    if (type !== undefined) params += (params ? '&' : '') + `type=${type}`;
    
    return this.http.get<Ad | null>(`${this.apiUrl}/serve?${params}`);
  }

  recordImpression(campaignId: number, videoId?: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/impressions`, {
      campaignId,
      videoId
    });
  }

  recordClick(campaignId: number, videoId?: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/clicks`, {
      campaignId,
      videoId
    });
  }

  // Analytics
  getAdRevenue(startDate?: string, endDate?: string): Observable<AdRevenue> {
    let params = '';
    if (startDate) params += `startDate=${startDate}`;
    if (endDate) params += (params ? '&' : '') + `endDate=${endDate}`;
    
    return this.http.get<AdRevenue>(`${this.apiUrl}/revenue?${params}`);
  }

  getCampaignAnalytics(campaignId: number, startDate?: string, endDate?: string): Observable<AdAnalytics> {
    let params = '';
    if (startDate) params += `startDate=${startDate}`;
    if (endDate) params += (params ? '&' : '') + `endDate=${endDate}`;
    
    return this.http.get<AdAnalytics>(`${this.apiUrl}/campaigns/${campaignId}/analytics?${params}`);
  }

  getAdvertiserAnalytics(startDate?: string, endDate?: string): Observable<AdAnalytics> {
    let params = '';
    if (startDate) params += `startDate=${startDate}`;
    if (endDate) params += (params ? '&' : '') + `endDate=${endDate}`;
    
    return this.http.get<AdAnalytics>(`${this.apiUrl}/analytics?${params}`);
  }
}

File: Backend/Drashyam.API/Services/AdService.cs
using AutoMapper;
using Drashyam.API.Data;
using Drashyam.API.DTOs;
using Drashyam.API.Models;
using Microsoft.EntityFrameworkCore;

namespace Drashyam.API.Services;

public class AdService : IAdService
{
    private readonly DrashyamDbContext _context;
    private readonly IMapper _mapper;
    private readonly ILogger<AdService> _logger;
    private readonly ISubscriptionService _subscriptionService;

    public AdService(DrashyamDbContext context, IMapper mapper, ILogger<AdService> logger, ISubscriptionService subscriptionService)
    {
        _context = context;
        _mapper = mapper;
        _logger = logger;
        _subscriptionService = subscriptionService;
    }

    public async Task<AdCampaignDto> CreateAdCampaignAsync(AdCampaignCreateDto createDto, string advertiserId)
    {
        var campaign = _mapper.Map<AdCampaign>(createDto);
        campaign.AdvertiserId = advertiserId;
        campaign.Status = AdStatus.Draft;
        campaign.CreatedAt = DateTime.UtcNow;

        _context.AdCampaigns.Add(campaign);
        await _context.SaveChangesAsync();

        return _mapper.Map<AdCampaignDto>(campaign);
    }

    public async Task<AdCampaignDto> GetAdCampaignByIdAsync(int campaignId)
    {
        var campaign = await _context.AdCampaigns
            .Include(a => a.Advertiser)
            .Include(a => a.Impressions)
            .FirstOrDefaultAsync(a => a.Id == campaignId);

        if (campaign == null)
            throw new ArgumentException("Ad campaign not found");

        return _mapper.Map<AdCampaignDto>(campaign);
    }

    public async Task<PagedResult<AdCampaignDto>> GetAdCampaignsAsync(string advertiserId, int page = 1, int pageSize = 20)
    {
        var campaigns = await _context.AdCampaigns
            .Where(a => a.AdvertiserId == advertiserId)
            .Include(a => a.Impressions)
            .OrderByDescending(a => a.CreatedAt)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        var totalCount = await _context.AdCampaigns
            .Where(a => a.AdvertiserId == advertiserId)
            .CountAsync();

        return new PagedResult<AdCampaignDto>
        {
            Items = _mapper.Map<List<AdCampaignDto>>(campaigns),
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<AdCampaignDto> UpdateAdCampaignAsync(int campaignId, AdCampaignUpdateDto updateDto, string advertiserId)
    {
        var campaign = await _context.AdCampaigns
            .FirstOrDefaultAsync(a => a.Id == campaignId && a.AdvertiserId == advertiserId);

        if (campaign == null)
            throw new ArgumentException("Ad campaign not found");

        _mapper.Map(updateDto, campaign);
        campaign.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();
        return _mapper.Map<AdCampaignDto>(campaign);
    }

    public async Task<bool> DeleteAdCampaignAsync(int campaignId, string advertiserId)
    {
        var campaign = await _context.AdCampaigns
            .FirstOrDefaultAsync(a => a.Id == campaignId && a.AdvertiserId == advertiserId);

        if (campaign == null)
            return false;

        _context.AdCampaigns.Remove(campaign);
        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<bool> ActivateAdCampaignAsync(int campaignId, string advertiserId)
    {
        var campaign = await _context.AdCampaigns
            .FirstOrDefaultAsync(a => a.Id == campaignId && a.AdvertiserId == advertiserId);

        if (campaign == null)
            return false;

        campaign.Status = AdStatus.Active;
        campaign.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<bool> PauseAdCampaignAsync(int campaignId, string advertiserId)
    {
        var campaign = await _context.AdCampaigns
            .FirstOrDefaultAsync(a => a.Id == campaignId && a.AdvertiserId == advertiserId);

        if (campaign == null)
            return false;

        campaign.Status = AdStatus.Paused;
        campaign.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<AdDto?> GetAdForUserAsync(string userId, int? videoId = null, Models.AdType? preferredType = null)
    {
        // Check if user has a paid subscription (no ads for paid users)
        try
        {
            var subscription = await _subscriptionService.GetUserSubscriptionAsync(userId);
            if (subscription.Plan != null && !subscription.Plan.HasAds)
            {
                return null; // No ads for paid users
            }
        }
        catch
        {
            // User has no subscription, show ads
        }

        // Get active campaigns that match criteria
        var query = _context.AdCampaigns
            .Where(a => a.Status == AdStatus.Active && 
                       a.StartDate <= DateTime.UtcNow && 
                       a.EndDate >= DateTime.UtcNow);

        if (preferredType.HasValue)
        {
            query = query.Where(a => a.Type == preferredType.Value);
        }

        var campaigns = await query.ToListAsync();

        if (!campaigns.Any())
            return null;

        // Simple random selection (in production, use more sophisticated targeting)
        var random = new Random();
        var selectedCampaign = campaigns[random.Next(campaigns.Count)];

        return new AdDto
        {
            Id = selectedCampaign.Id,
            Type = (DTOs.AdType)selectedCampaign.Type,
            Content = selectedCampaign.AdContent,
            Url = selectedCampaign.AdUrl,
            ThumbnailUrl = selectedCampaign.ThumbnailUrl,
            CostPerClick = selectedCampaign.CostPerClick,
            CostPerView = selectedCampaign.CostPerView
        };
    }

    public async Task<bool> RecordAdImpressionAsync(int campaignId, string userId, int? videoId = null)
    {
        var campaign = await _context.AdCampaigns.FindAsync(campaignId);
        if (campaign == null)
            return false;

        var impression = new AdImpression
        {
            AdCampaignId = campaignId,
            UserId = userId,
            VideoId = videoId,
            ViewedAt = DateTime.UtcNow,
            Revenue = campaign.CostPerView
        };

        _context.AdImpressions.Add(impression);
        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<bool> RecordAdClickAsync(int campaignId, string userId, int? videoId = null)
    {
        var impression = await _context.AdImpressions
            .FirstOrDefaultAsync(i => i.AdCampaignId == campaignId && 
                                     i.UserId == userId && 
                                     i.VideoId == videoId);

        if (impression == null)
            return false;

        impression.WasClicked = true;
        impression.ClickedAt = DateTime.UtcNow;

        var campaign = await _context.AdCampaigns.FindAsync(campaignId);
        if (campaign != null)
        {
            impression.Revenue = campaign.CostPerClick;
        }

        await _context.SaveChangesAsync();
        return true;
    }

    public async Task<decimal> GetAdRevenueAsync(string userId, DateTime? startDate = null, DateTime? endDate = null)
    {
        var query = _context.AdImpressions
            .Where(i => i.UserId == userId);

        if (startDate.HasValue)
            query = query.Where(i => i.ViewedAt >= startDate.Value);

        if (endDate.HasValue)
            query = query.Where(i => i.ViewedAt <= endDate.Value);

        return await query.SumAsync(i => i.Revenue);
    }

    public async Task<AdRevenueDto> GetAdRevenueBreakdownAsync(string userId, DateTime? startDate = null, DateTime? endDate = null)
    {
        var query = _context.AdImpressions
            .Where(i => i.UserId == userId);

        if (startDate.HasValue)
            query = query.Where(i => i.ViewedAt >= startDate.Value);

        if (endDate.HasValue)
            query = query.Where(i => i.ViewedAt <= endDate.Value);

        var impressions = await query.Include(i => i.AdCampaign).ToListAsync();

        var totalRevenue = impressions.Sum(i => i.Revenue);
        var totalImpressions = impressions.Count;
        var totalClicks = impressions.Count(i => i.WasClicked);

        return new AdRevenueDto
        {
            TotalRevenue = totalRevenue,
            TotalImpressions = totalImpressions,
            TotalClicks = totalClicks,
            ClickThroughRate = totalImpressions > 0 ? (decimal)totalClicks / totalImpressions : 0,
            RevenuePerImpression = totalImpressions > 0 ? totalRevenue / totalImpressions : 0,
            RevenuePerClick = totalClicks > 0 ? totalRevenue / totalClicks : 0
        };
    }

    public async Task<PagedResult<AdImpressionDto>> GetAdImpressionsAsync(int campaignId, int page = 1, int pageSize = 20)
    {
        var impressions = await _context.AdImpressions
            .Where(i => i.AdCampaignId == campaignId)
            .Include(i => i.User)
            .Include(i => i.Video)
            .OrderByDescending(i => i.ViewedAt)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        var totalCount = await _context.AdImpressions
            .Where(i => i.AdCampaignId == campaignId)
            .CountAsync();

        return new PagedResult<AdImpressionDto>
        {
            Items = _mapper.Map<List<AdImpressionDto>>(impressions),
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<AdAnalyticsDto> GetAdCampaignAnalyticsAsync(int campaignId, DateTime? startDate = null, DateTime? endDate = null)
    {
        var query = _context.AdImpressions
            .Where(i => i.AdCampaignId == campaignId);

        if (startDate.HasValue)
            query = query.Where(i => i.ViewedAt >= startDate.Value);

        if (endDate.HasValue)
            query = query.Where(i => i.ViewedAt <= endDate.Value);

        var impressions = await query.ToListAsync();

        var totalImpressions = impressions.Count;
        var totalClicks = impressions.Count(i => i.WasClicked);
        var totalRevenue = impressions.Sum(i => i.Revenue);

        return new AdAnalyticsDto
        {
            CampaignId = campaignId,
            TotalImpressions = totalImpressions,
            TotalClicks = totalClicks,
            TotalRevenue = totalRevenue,
            ClickThroughRate = totalImpressions > 0 ? (decimal)totalClicks / totalImpressions : 0,
            CostPerClick = totalClicks > 0 ? totalRevenue / totalClicks : 0,
            CostPerImpression = totalImpressions > 0 ? totalRevenue / totalImpressions : 0
        };
    }

    public async Task<AdAnalyticsDto> GetAdvertiserAnalyticsAsync(string advertiserId, DateTime? startDate = null, DateTime? endDate = null)
    {
        var baseQuery = _context.AdImpressions
            .Include(i => i.AdCampaign)
            .Where(i => i.AdCampaign.AdvertiserId == advertiserId);

        if (startDate.HasValue)
            baseQuery = baseQuery.Where(i => i.ViewedAt >= startDate.Value);

        if (endDate.HasValue)
            baseQuery = baseQuery.Where(i => i.ViewedAt <= endDate.Value);

        var impressions = await baseQuery.ToListAsync();

        var totalImpressions = impressions.Count;
        var totalClicks = impressions.Count(i => i.WasClicked);
        var totalRevenue = impressions.Sum(i => i.Revenue);

        return new AdAnalyticsDto
        {
            CampaignId = 0, // Not applicable for advertiser analytics
            TotalImpressions = totalImpressions,
            TotalClicks = totalClicks,
            TotalRevenue = totalRevenue,
            ClickThroughRate = totalImpressions > 0 ? (decimal)totalClicks / totalImpressions : 0,
            CostPerClick = totalClicks > 0 ? totalRevenue / totalClicks : 0,
            CostPerImpression = totalImpressions > 0 ? totalRevenue / totalImpressions : 0
        };
    }
}